### ISSUE TITLE
Global and Point Features use for 3D BoNet

### ISSUE THREAD

**Issue by escanillans** [QingyongHu/RandLA-Net#128]
Hi @Yang7879. This is a great paper!

I was wondering if you could help me - Is it possible to extract the global and point features to be used as the backbone for 3D BoNet? If so, where would I do that? Thanks

**Comment by Yang7879** [comment]
Hi @escanillans, RandLA-Net is very suitable to be a more powerful backbone for 3D-BoNet.  You may add few more MLP layers to get a global vector in the middle of RandLA, and at the end of RandLA to get per-point feature vectors.

**Comment by escanillans** [comment]
Thank you for the quick reply @Yang7879. I am currently trying to decide where to place more MLP layers but wanted to check with you first if this is also what you had in mind:

`    def building_block(self, xyz, feature, neigh_idx, d_out, name, is_training):
        d_in = feature.get_shape()[-1].value
        f_xyz = self.relative_pos_encoding(xyz, neigh_idx)
        f_xyz = helper_tf_util.conv2d(f_xyz, d_in, [1, 1], name + 'mlp1', [1, 1], 'VALID', True, is_training)
        f_neighbours = self.gather_neighbour(tf.squeeze(feature, axis=2), neigh_idx)
        f_concat = tf.concat([f_neighbours, f_xyz], axis=-1)
        f_pc_agg = self.att_pooling(f_concat, d_out // 2, name + 'att_pooling_1', is_training)

        # features_global = helper_tf_util.conv2d(...)

        f_xyz = helper_tf_util.conv2d(f_xyz, d_out // 2, [1, 1], name + 'mlp2', [1, 1], 'VALID', True, is_training)
        f_neighbours = self.gather_neighbour(tf.squeeze(f_pc_agg, axis=2), neigh_idx)
        f_concat = tf.concat([f_neighbours, f_xyz], axis=-1)
        f_pc_agg = self.att_pooling(f_concat, d_out, name + 'att_pooling_2', is_training)

        # features_point = helper_tf_util.conv2d(...)

        return f_pc_agg`

Of course in the `return` statement I would return the point and global features as well and pass that to 3D BoNet. However, I'm still trying to figure out if this would be the correct placement.

**Comment by escanillans** [comment]
Hi @Yang7879,

I am now able to output the global and point features for use in 3D BoNet. However, I was wondering if you could give some guidance as to how to prepare the dataset. For example, using S3DIS, the formats used in RandLANet and BoNet are quite different (RandLANet has the .pkl and .ply files whereas BoNet contains .h5 files). Thanks

**Comment by Katherine2013** [comment]
> Thank you for the quick reply @Yang7879. I am currently trying to decide where to place more MLP layers but wanted to check with you first if this is also what you had in mind. I'm still trying to figure out if this would be the correct placement but it would be great to have your advice on this!
> 
> ```
>     def inference(self, inputs, is_training):
> 
>         d_out = self.config.d_out
>         feature = inputs['features']
>         feature = tf.layers.dense(feature, 8, activation=None, name='fc0')
>         feature = tf.nn.leaky_relu(tf.layers.batch_normalization(feature, -1, 0.99, 1e-6, training=is_training))
>         feature = tf.expand_dims(feature, axis=2)
> 
>         # ###########################Encoder############################
>         f_encoder_list = []
>         for i in range(self.config.num_layers):
>             f_encoder_i = self.dilated_res_block(feature, inputs['xyz'][i], inputs['neigh_idx'][i], d_out[i],
>                                                  'Encoder_layer_' + str(i), is_training)
>             f_sampled_i = self.random_sample(f_encoder_i, inputs['sub_idx'][i])
>             feature = f_sampled_i
> 
>             # midway through training
>             # features_global = helper_tf_util.conv2d(...)
> 
>             # towards end of config.num_layers
>             # features_point = helper_tf_util.conv2d(...)
> 
>             if i == 0:
>                 f_encoder_list.append(f_encoder_i)
>             f_encoder_list.append(f_sampled_i)
>         # ###########################Encoder############################
> 
>         feature = helper_tf_util.conv2d(f_encoder_list[-1], f_encoder_list[-1].get_shape()[3].value, [1, 1],
>                                         'decoder_0',
>                                         [1, 1], 'VALID', True, is_training)
> 
>         # ###########################Decoder############################
>         f_decoder_list = []
>         for j in range(self.config.num_layers):
>             f_interp_i = self.nearest_interpolation(feature, inputs['interp_idx'][-j - 1])
>             f_decoder_i = helper_tf_util.conv2d_transpose(tf.concat([f_encoder_list[-j - 2], f_interp_i], axis=3),
>                                                           f_encoder_list[-j - 2].get_shape()[-1].value, [1, 1],
>                                                           'Decoder_layer_' + str(j), [1, 1], 'VALID', bn=True,
>                                                           is_training=is_training)
>             feature = f_decoder_i
>             f_decoder_list.append(f_decoder_i)
>         # ###########################Decoder############################
> 
>         f_layer_fc1 = helper_tf_util.conv2d(f_decoder_list[-1], 64, [1, 1], 'fc1', [1, 1], 'VALID', True, is_training)
>         f_layer_fc2 = helper_tf_util.conv2d(f_layer_fc1, 32, [1, 1], 'fc2', [1, 1], 'VALID', True, is_training)
>         f_layer_drop = helper_tf_util.dropout(f_layer_fc2, keep_prob=0.5, is_training=is_training, scope='dp1')
>         f_layer_fc3 = helper_tf_util.conv2d(f_layer_drop, self.config.num_classes, [1, 1], 'fc', [1, 1], 'VALID', False,
>                                             is_training, activation_fn=None)
>         f_out = tf.squeeze(f_layer_fc3, [2])
>         return f_out
> ```

Hi @escanillans I'm also working on integrating RandLA into BoNet. Is this the correct placement? Looking forward to your reply!
